adapter/src/ethereum.rs

```diff
@@ -113,95 +110,103 @@ impl Adapter for EthereumAdapter {
         &self.address
     }
 
-    fn sign(&self, state_root: &str) -> AdapterResult<String> {
+    fn sign(&self, state_root: &str) -> AdapterResult<String, Self::AdapterError> {
         if let Some(wallet) = &self.wallet {
-            let state_root = hex::decode(state_root)
-                .map_err(|_| AdapterError::Signature("invalid state_root".to_string()))?;
+            let state_root =
+                hex::decode(state_root).map_err(StateRootError::StateRootHexDecoding)?;
             let message = Message::from_slice(&hash_message(&state_root));
             let wallet_sign = wallet
                 .sign(&self.keystore_pwd, &message)
-                .map_err(|_| map_error("failed to sign messages"))?;
+                .map_err(EwtSigningError::SigningMessage)?;
             let signature: Signature = wallet_sign.into_electrum().into();
 
             Ok(format!("0x{}", signature))
         } else {
-            Err(AdapterError::Configuration(
-                "Unlock the wallet before signing".to_string(),
-            ))
+            Err(AdapterError::LockedWallet)
         }
     }
 
-    fn verify(&self, signer: &ValidatorId, state_root: &str, sig: &str) -> AdapterResult<bool> {
+    fn verify(
+        &self,
+        signer: &ValidatorId,
+        state_root: &str,
+        sig: &str,
+    ) -> AdapterResult<bool, Self::AdapterError> {
         if !sig.starts_with("0x") {
-            return Err(AdapterError::Signature("not 0x prefixed hex".to_string()));
+            return Err(StateRootError::SignatureNotPrefixed.into());
```

Comment:
Signature errors aren't a type of StateRootError it's quite different. Signature errors could either standalone or may be part of some struct with a more suitable name context

Reply:
Yes, we need to find a different name for this error, but in a nutshell we want to specify that that decoding the `signature`, `state_root` or verifying the `address, signature, message` failed. Another way is to completely remov this error by accepting something like `Vec<u8>/[u8]`.
This can allow us to still have `String/&str` for DummyAdapter & remove all the `decode`, `starts_with("0x")`, etc.